mne/preprocessing/faster.py

```diff
@@ -0,0 +1,463 @@
+                          "the tested frequencies for this metric or use the "
+                          "'metrics' parameter to disable the "
+                          "'power_gradient' metric.").format(prange[0]))
+    ps = ps[:, start:stop]
+
+    # Compute mean gradients
+    return np.mean(np.diff(ps), axis=1)
+
+
+def _deviation(data):
+    """Computes the deviation from mean for each channel in a set of epochs.
+
+    This is not implemented as a lambda function, because the channel means
+    should be cached during the computation.
+
+    Parameters
+    ----------
+    data : 3D numpy array
+        The epochs (#epochs x #channels x #samples).
+
+    Returns
+    -------
+    dev : 1D numpy array
+        For each epoch, the mean deviation of the channels.
+    """
+    ch_mean = np.mean(data, axis=2)
+    return ch_mean - np.mean(ch_mean, axis=0)
+
+
+def faster_bad_channels(epochs, picks=None, thres=3, use_metrics=None,
+                        max_iter=2, return_by_metric=False):
+    """Implements the first step of the FASTER algorithm.
+
+    This function attempts to automatically mark bad EEG channels by performing
+    outlier detection. It operated on epoched data, to make sure only relevant
+    data is analyzed.
+
+    Parameters
+    ----------
+    epochs : Instance of Epochs
+        The epochs for which bad channels need to be marked
+    picks : list of int | None
+        Channels to operate on. Defaults to EEG channels.
+    thres : float
+        The threshold value, in standard deviations, to apply. A channel
+        crossing this threshold value is marked as bad. Defaults to 3.
+    use_metrics : list of str
+        List of metrics to use. Can be any combination of:
+            'variance', 'correlation', 'hurst', 'kurtosis', 'line_noise'
+        Defaults to all of them.
+    max_iter : int
+        The maximum number of iterations performed during outlier detection
+        (defaults to 2, original FASTER paper uses 1).
+    return_by_metric : bool
+        Whether to return the bad channels as a flat list (False, default) or
+        as a dictionary with the names of the used metrics as keys and the
+        bad channels found by this metric as values.
+
+    Returns
+    -------
+    bads : list of str
+        The names of the bad EEG channels.
+    """
+    metrics = {
+        'variance': lambda x: np.var(x, axis=1),
+        'correlation': lambda x: np.mean(
+            np.ma.masked_array(np.corrcoef(x),
+                               np.identity(len(x), dtype=bool)), axis=0),
+        'hurst': lambda x: _hurst(x),
+        'kurtosis': lambda x: kurtosis(x, axis=1),
+        'line_noise': lambda x: _freqs_power(x, epochs.info['sfreq'],
+                                             [50, 60]),
+    }
+
+    if picks is None:
+        picks = pick_types(epochs.info, meg=True, eeg=True, exclude=[])
+    if use_metrics is None:
+        use_metrics = metrics.keys()
+
+    # Concatenate epochs in time
```

Comment:
this is also done in `decoding.classifier.ConcatenateChannels`. Maybe we need to refactor the common components ...